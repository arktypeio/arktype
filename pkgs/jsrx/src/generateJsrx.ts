import { readJsonSync, writeFileSync } from "fs-extra"
import { join } from "path"
import { prompt, shell } from "@re-do/node-utils"

export type GenerateJsrxOptions = {
    language: "js" | "ts"
}

const jsImportsTemplate = `const { jsrx, $, shell } = require("jsrx")`

const tsImportsTemplate = `import { jsrx, $, shell } from "jsrx"`

export const generateFileContents = (
    scripts: Record<string, string>,
    { language }: GenerateJsrxOptions
) => `${language === "js" ? jsImportsTemplate : tsImportsTemplate}

jsrx({
    dev: {
        ${Object.entries(scripts)
            .map(([name, command]) => {
                if (command.startsWith("jsrx")) {
                    throw new Error(
                        "Cannot generate a jsrx file from package.json scripts that already rely on jsrx. If you've deleted your jsrx file, you should manually put it back."
                    )
                }
                return `${
                    // Put script name in quotes if it's not alphanumeric
                    /[^a-zA-Z0-9]/.test(name) ? `"${name}"` : name
                    // Put script value in backticks to try and avoid conflicts
                }: $(\`${command}\`)`
            })
            .join(",\n        ")}
    },
    prod: {},
    shared: {}
})
`

export const generateJsrx = (options: GenerateJsrxOptions) => {
    const packageJsonFile = join(process.cwd(), "package.json")
    const packageJsonContents = readJsonSync(packageJsonFile)
    const scripts = packageJsonContents.scripts ?? {}
    const fileToGenerate = join(
        process.cwd(),
        `jsrx.${options.language === "js" ? "js" : "ts"}`
    )
    writeFileSync(fileToGenerate, generateFileContents(scripts, options))
    return fileToGenerate
}

export const promptForJsrxFile = async () => {
    const shouldGenerate = await prompt(
        `Found no 'jsrx.js' or 'jsrx.ts' file in ${process.cwd()}. Would you like to generate one from your package.json?`,
        "confirm"
    )
    if (shouldGenerate) {
        const language = await prompt(
            `Which language would you like to generate?`,
            "select",
            {
                choices: [
                    { title: "JavaScript (jsrx.js)", value: "js" },
                    { title: "TypeScript (jsrx.ts)", value: "ts" }
                ]
            }
        )
        const generatedFile = generateJsrx({ language })
        const shouldUpdatePackageJson = await prompt(
            "Would you like to automatically update your package.json scripts to use jsrx?",
            "confirm"
        )
        if (shouldUpdatePackageJson) {
            shell("jsrx jsrxGen")
        }
        return generatedFile
    } else {
        throw Error("A jsrx file is required.")
    }
}
