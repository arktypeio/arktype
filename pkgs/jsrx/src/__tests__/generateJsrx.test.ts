import { generateFileContents } from "../generateJsrx"

const scripts = {
    sayHello: "echo hello",
    sayGoodBye: "echo goodBye",
    "handle:non-alpha": "echo handled"
}

it("generates js", () => {
    expect(generateFileContents(scripts, { language: "js" })).toMatchSnapshot()
})

it("generates ts", () => {
    expect(generateFileContents(scripts, { language: "ts" })).toMatchSnapshot()
})
