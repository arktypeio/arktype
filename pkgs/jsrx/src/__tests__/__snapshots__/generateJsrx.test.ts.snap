// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates js 1`] = `
"const { jsrx, $, shell } = require(\\"jsrx\\")

jsrx({
    dev: {
        sayHello: $(\`echo hello\`),
        sayGoodBye: $(\`echo goodBye\`),
        \\"handle:non-alpha\\": $(\`echo handled\`)
    },
    prod: {},
    shared: {}
})
"
`;

exports[`generates ts 1`] = `
"import { jsrx, $, shell } from \\"jsrx\\"

jsrx({
    dev: {
        sayHello: $(\`echo hello\`),
        sayGoodBye: $(\`echo goodBye\`),
        \\"handle:non-alpha\\": $(\`echo handled\`)
    },
    prod: {},
    shared: {}
})
"
`;
